### PR TITLE
Redis memory usage optimizations

### LINKED ISSUE #7: Redis garbage collection
Currently emitted ops are retained in redis in two separate keys until the TTL expires:

- ``op:${opId}``
- ``emittedEvents:${opId}``

To reduce redis memory usage, there should be a background process that deletes the `op` key if the corresponding `emittedEvents` key exists.

Trying to clean up these keys in real-time is not possible due to race conditions with redundant processes.

### PR DESCRIPTION
**Pull request by will123195**
resolves #7

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -300,18 +303,29 @@ class MongoOplogReader {
       });
   }
 
+  deleteOpMembers(opId) {
+    // wait a few seconds to prevent redundant workers from recreating the key immediately after it's deleted
+    // if a redundant worker is way slower, this key will be purged passively after the TTL
+    setTimeout(() => {
```
Comment (aoxrud) [inline src/index.js:309]: I don't like the `setTimeout` solution because it assumes that whatever might be happening is done in `N seconds` and that may not always be true.

Are there any other solutions?

Food for thought: Would be nice if each `op` could set a "clean up" flag. 

(Maybe I am misunderstanding the entire purpose of the code, if so, please explain)

Comment (will123195) [reply src/index.js:309]: Agreed, I like to avoid setTimeout. But the only alternative I can think of is to write a garbage collector that runs in the background--which is less effective/efficient.

Here's the problem, we have redundant workers receiving "oplog events". As soon as two members have seen the oplog event, we emit the "food event" and then we need to delete `members`.

Suppose workers receive the oplog event in this order:
```
worker  members
A1      [A]
B1      [A,B]  --> we have 2 members! emit the food event and delete members
A2      [A]
B2      [A,B]  --> we have 2 members! (try to) emit the food event and delete members
```
In this case we're all good because `members` has been deleted.

But suppose workers receive the oplog event in this order:
```
worker  members
A1      [A]
A2      [A]
B1      [A,B]  --> we have 2 members! emit the event and delete members
B2      [B]
```

You can see B2 fires a few milliseconds after `members` is deleted and recreates `members`. 

The obvious fix is `setTimeout` but I'm open to alternatives. I agree `N seconds` is arbitrary but I set it high enough that I'm not concerned about the outliers that may not get deleted.

**Review by aoxrud — APPROVED** [review]
Thank you for the detailed explanation. I don't think that having a GC background process is "inefficient" -- I think it is probably the proper way to solve for this issue.

However, it is likely an edge case scenario that may not be worth the time spent to solve it. @wwbenomathew thoughts?

I think the proposed solution is probably good enough, maybe lets give this simple solution a test and if/when we detect that things are slipping through the cracks, we can use the more involved solution. (GC background process)

File: src/index.js [hunk 2]
```diff
@@ -300,18 +303,29 @@ class MongoOplogReader {
...
+      const key = this.getOpMembersKey(opId);
+      this.redisClient.delAsync(key);
+    }, 3000);
+  }
+
   processOp(data, replSetName, memberName) {
     debug(`processOp: ${replSetName} ${memberName} %o`, data);
-    return this.isOpMajorityDetected(replSetName, memberName, data)
+    const opId = this.getOpId(data);
+    return this.isOpMajorityDetected(replSetName, memberName, opId)
```
Comment (wwbenomathew) [inline src/index.js:318]: what if `isOpMajorityDetected` happens only if `opId` is not in the in emitted events key/set ? we might be able to avoid the set timeout and instead just have a 24 hour TTL on the `${this.keyPrefix}:op:${opId}` key to handle corner cases.

Comment (will123195) [reply src/index.js:318]: I don't think that will work because, in theory, all processes will be making this check at the same time before the event is emitted.

Comment (wwbenomathew) [reply src/index.js:318]: Im wondering if this check is a real world scenario since Redis is [single threaded ](https://redis.io/topics/faq)- section `Redis is single threaded. How can I exploit multiple CPU / cores?`. Also, another thing that can be considered is [redis transactions ](https://redis.io/topics/transactions) since transactions are executed atomically.

Comment (will123195) [reply src/index.js:318]: Despite being single threaded, I think redis still has async i/o. 
https://stackoverflow.com/questions/10489298/redis-is-single-threaded-then-how-does-it-do-concurrent-i-o

Also I don't think transactions work here since the multiple commands are issued on different client connections.

Comment (wwbenomathew) [reply src/index.js:318]: Ok, that makes sense.

**Review by wwbenomathew — APPROVED** [review]
